### PR TITLE
Automatically generate use case files through Caze

### PR DESCRIPTION
**Pull request by luizParreira**
This __PR__ implements a first version of a generator that could generate files dynamically by running:

```
$ bundle exec caze generate engine:my_engine:my_use_case_module:my_class
```

The command above will create 2 new files:

```
# ->  engines/my_engine/lib/my_engine/my_use_case_module/my_class.rb
# ->  engines/my_engine/spec/my_engine/my_use_case_module/my_class_spec.rb
```

Please note, that I still need to write some simple documentation for this. My intent on opening this PR was to allow reviewers to start reviewing the code upfront.

Also, additionally this adds a sample `Dockerfile` and `docker-compose.yml`, so that those using `docker-machine` can use this gem using docker.

Any ideas on how to improve the code are welcome. 

TODO:
- [x] Write doc for this generator
- [ ] Write `help` method
- [ ] Add the init params as the commands second params
- [ ] Add aliases to the command (g for generate)?
- [ ] Populate files with template code
- [x] Increase test coverage
- [ ] Better error handling, what happens when a file already exists, and if a path doesnt exist?

### CODE REVIEW

**Comment by luizParreira** [comment]
@philss I think we should be able to merge this already, and I can add the rest of the bullet points highlighted here on another PR, what do you think?

File: lib/caze/generator.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  # Public: Module responsible for generating use case files
+  module Generator
+    autoload :CLI, 'caze/generator/cli'
+
+    class << self
+      def start
+        CLI.start
+      end
+
+      def known_path?(path_type)
```
Comment (philss) [inline lib/caze/generator.rb]: can we have `lib/` here too?

File: lib/caze/generator/parser.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          :app
+        end
+
+        def resolve_dependency_namespace(namespaces, dependency_type)
+          return namespaces.delete_at(0) if known_path? dependency_type
+          :main_app
+        end
+
+        def resolve_internal_modules(namespaces)
+          n = namespaces.count
```
Comment (philss) [inline lib/caze/generator/parser.rb]: can we rename `n` to `count`?

File: lib/caze/generator/parser.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+          return '' if n <= 1
+
+          modules = ''
+          while n > 1
+            modules += "#{namespaces.delete_at 0}/"
```
Comment (philss) [inline lib/caze/generator/parser.rb]: you can `namespaces.shift` instead of `delete_at 0`.

File: lib/caze/generator/parser.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            n = namespaces.count
+          end
+          modules
+        end
+
+        def pluralize(path_type)
+          return "#{path_type}s".to_sym if [:gem, :engine, :module].include? path_type
```
Comment (philss) [inline lib/caze/generator/parser.rb]: can we add that list to a constant? also, you could: `%i(gem engine module)` to produce the same list

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -14,6 +14,9 @@
 # users commonly want.
 #
 # See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
+
+require 'byebug'
```
Comment (philss) [inline spec/spec_helper.rb]: is this necessary here?

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ENV PATH $GEM_HOME/bin:$PATH
+ENV HISTFILE $APP_HOME/tmp/docker_histfile
+ENV LANG C.UTF-8
+
+# Config for the terminal
+ADD .bashrc /root/.bashrc
+
+RUN bundle check || bundle install -j4
+
+CMD bundle exec rspec spec --color
```
Comment (philss) [inline Dockerfile]: since we added a Docker/Docker compose structure, would be awesome to add instructions to our README.md on how to get things running in development. What do you think?

This could be another PR.

**Comment by lucasprag** [comment]
What an awesome idea! Keep the good work guys @luizParreira @philss =)
